### ISSUE TITLE
Implement over Node Media Server

### ISSUE THREAD

**Issue by carlosanpardo** [johndavedecano/node-rtmp-hls#1]
Hello Jhon. I copied the configuration of the file "[https://github.com/johndavedecano/node-rtmp-hls/blob/master/src/config.js](https://github.com/johndavedecano/node-rtmp-hls/blob/master/src/config.js)" of your project in my Node Media Server and the server created one folder per each trans-tasks but When I open each m3u8 file the resolution is the same in all profiles. 

I suspect that the encoding video process is not working.

How I can implement your files in a fresh Node Media Server? Thank you for your help.

**Comment by carlosanpardo** [comment]
Well, I copied the files into the Node Media Server and execute the command "node index.js" and the server start (check the ffmpeg path). 

The main index.m3u8 file was created with the path for the another resolutions too. The issue is that  the encoding video process is not working. The output quality is the same in all profiles:

hls_360p
hls_480p
hls_720p
hls_1080p

The audio encoding is working fine. 

Can you give me a hint about this issue?

 I appreciate your help.

**Comment by johndavedecano** [comment]
Most like the ffmpeg itself. Let me check on that

**Comment by carlosanpardo** [comment]
Thank you. Any advance over this issue? I reinstalled the Node Media Server with in the last version and place your files and now appear an error with mkdirp. 

Can you have a full copy  of the project? 
Can you share me details about OS and FFMPEG Version? I used Centos 7.X and Ubuntu 16.04 and the same errors in both OS.

I appreciate your help.

**Comment by johndavedecano** [comment]
I just made some updates check it out

**Comment by carlosanpardo** [comment]
Hello John. Thank you for your prompt help!

Bellow my results:

1) I compiled the project and the server started without issues. Including the FFMPEG.
2) In my OBS the app "live" not created the HLS files. I changed the name of the app for "stream" and the m3u8 files was created (rtmp://51.158.120.90:1935/stream). Please check an "if clausule" that you have in server.js from the line "21".
3) The main m3u8 file have this path: http://51.158.120.90:8000/live/expressmedia.m3u8. Into the main documentation of the project appear a wrong URL "http://localhost:8000/live/STREAM_NAME/index.m3u8".
4) The output resolutions is the same in all output files:
- hls_1080p
- hls_720p
- hls_480p
- hls_360p

Please check the results of the htop command that show that the video parameters are not passed [https://snipboard.io/kAUv7g.jpg](https://snipboard.io/kAUv7g.jpg)

I appreciate your help with this issue and your dedication. Please ping me if need anything from my side. Thank you.

**Comment by donatoferrulli** [comment]
HTH vcParams should be vcParam.

**Comment by carlosanpardo** [comment]
> HTH vcParams should be vcParam.

Hello @donatoferrulli . Thank you for your message. I changed "vcParams" to "vcParam" and the additionals m3u8 files was not created. 

```
./hls_360p/expressmedia/index.m3u8
./hls_480p/expressmedia/index.m3u8
./hls_720p/expressmedia/index.m3u8
./hls_1080p/expressmedia/index.m3u8
```

In other words, the "vcParams" is correct!

**Comment by donatoferrulli** [comment]
As you can see here https://github.com/illuspas/Node-Media-Server/blob/master/node_trans_session.js

The correct parameter is vcParam. I'm not saying that this will fix the project but with htop you will see the setted vcParam.

Now we should try to set the correct parameters which i don't know.

@johndavedecano any suggestions?

**Comment by lofar788** [comment]
@carlosanpardo have you found a solution? I'm having the same problem

**Comment by carlosanpardo** [comment]
@lofar788 Hello. I can't use the transcode but I have the idea of pass the parameter to encode to another resolution an replace into the line 60 an remove the line 63 in the file https://github.com/illuspas/Node-Media-Server/blob/master/node_trans_session.js

Can you try?

Currently in my nginx i transcode the video using this code and works fine: 

        exec ffmpeg -i rtmp://127.0.0.1/app/$name
              -c:a libfdk_aac -b:a 128k -c:v libx264 -b:v 1280k -f flv -g 30 -r 30 -s 1280x720 -preset medium -profile:v baseline rtmp://127.0.0.1/streamer/$name_720p2628kbs
              #-c:a libfdk_aac -b:a 128k -c:v libx264 -b:v 1000k -f flv -g 30 -r 30 -s 854x480 -preset veryfast -profile:v baseline rtmp://127.0.0.1/streamer/$name_480p1128kbs
              -c:a libfdk_aac -b:a 96k -c:v libx264 -b:v 750k -f flv -g 30 -r 30 -s 640x360 -preset veryfast -profile:v baseline rtmp://127.0.0.1/streamer/$name_360p878kbs
              #-c:a libfdk_aac -b:a 128k -c:v libx264 -b:v 400k -f flv -g 30 -r 30 -s 426x240 -preset veryfast -profile:v baseline rtmp://127.0.0.1/streamer/$name_240p528kbs
              -c:a libfdk_aac -b:a 64k -c:v libx264 -b:v 200k -f flv -g 15 -r 15 -s 426x240 -preset veryfast -profile:v baseline rtmp://127.0.0.1/streamer/$name_240p264kbs;



Thank you.

**Comment by DaveShearing** [comment]
@lofar788 @carlosanpardo I got this to work by adding in `vc: 'libx264',` above vcParam :)

**Comment by iabtyagi** [comment]
After struggling with the same issue for a few days and trying out various ffmpeg options, here is what I have found. Hope it helps others.

1. It's `**vcParam**` and not `vcParams`. It was discussed above but adding it to summarize. The reason vcParams appears to work is because it gets ignored and the library uses default options.
2. The scale value in the config and in many other examples elsewhere is within two pairs of quotes. That needs to be changed. `"'scale=1280:-1'"` --> `'scale=1280:-1'`
3. It is better to try out -2 instead of -1 as some codecs expect that. (Oversimplified statement, but -2 should work)
4. I found it better to have an additional param `-g` with a value about double of frame rate of source video so that the .ts files generated had proper length of 2 seconds (hls_time value in my case). My input source was 15 fps, so I set -g to 30.

So here is the config that worked for me:
```
vc: 'libx264',
vcParam: ['-g', '30', '-vf', 'scale=1280:-2', '-b:v', '2800k', '-preset', 'fast', '-profile:v', 'baseline', '-bufsize', '4200k'],
```

About the scale filter: https://trac.ffmpeg.org/wiki/Scaling
